hermit-abi/src/lib.rs

```diff
@@ -11,7 +11,15 @@ pub mod tcpstream;
 use libc::c_void;
 
 // sysmbols, which are part of the library operating system
+
+extern "Rust" {
+	fn sys_secure_rand64() -> Option<u64>;
+	fn sys_secure_rand32() -> Option<u32>;
+}
```

Comment:
So having `extern "Rust"` here means that the version of `rustc` used to compile `hermit-abi` must be _exactly_ the same version used to compile the kernel (`rusty-hermit`/`libhermit-rs`). Is this a problem? (I don't know much about HermitCore).

If it is, you could just have a returned value of `0` represent failure (I realize this is technically a valid random number, but it's extremely unlikely that it gets returned). Alternatively, you could take a `*mut u64` as input and return `0`/`-ENOSYS` on success/failure.

Reply:
In principle, the unikernel is completely written in Rust. Technical it could be possible that the Rust standard library call directly the Rust function of the unikernel. However, the unikernel depends on too many crates. The Rust community will not permit that rustc depends on all these crates. The workaround is a small system interface `hermit-abi`, which provides a similar interface like the crate `libc`. The unikernel itself will be provided as common static library. To guarantee that the unikernel and the application are build by the same compiler, we provide the crate `hermit-sys`. This crate builds the unikernel and link the library to the application. This is the reason that we suggest to add following line to `Cargo.toml`:

> [target.'cfg(target_os = "hermit")'.dependencies]
> hermit-sys = "0.1.*"

Unfortunately, every jump into a static library is `unsafe` because the compiler is not able to check the function. But we are able to use "real" Rust functions and to use `Option` as return value.